litmus/libdir/_aarch64/self.c

```diff
@@ -43,3 +43,17 @@ static uint32_t getcachelinesize(void) {
   uint32_t x = (uint32_t) (sz1 >= sz2 ? sz1 : sz2);
   return (1 << x) * 4 ;
 }
+
+inline static void check_dic_idc(uint8_t need_dic, uint8_t need_idc) {
+  uint64_t ctr_el0;
+  asm volatile ("mrs %x0, ctr_el0" : "=r" (ctr_el0));
+  need_idc = need_idc? 1 : 0;
+  need_dic = need_dic? 1 : 0;
+  uint64_t test_idc = (ctr_el0 & ((uint64_t)1) << 28) ^ ((uint64_t)need_idc) << 28;
+  uint64_t test_dic = (ctr_el0 & ((uint64_t)1) << 29) ^ ((uint64_t)need_dic) << 29;
+  if (test_idc || test_dic) {
+    fprintf(stderr, "Fatal error: hardware does not support the CacheType features requested by the litmus test\n");
```

Comment:
```suggestion
  bool idc = (ctr_el0 >> 28) & 1;
  bool dic = (ctr_el0 >> 29) & 1;
  if ((need_dic && !dic) || (need_idc && !idc)) {
    fprintf(stderr, "Fatal error: hardware does not support the CacheType "
            "feature IDC=%d, DIC=%d\n", need_idc, need_dic);
```

Reply:
Thank you, this is much tidier. Applying.